[css-scroll-snap] focus should not take into account scroll margin for visibility check

nmoucht opened the pull request:
#### f121c13339f2b359b9a6aaf8e9186ec8c539f58e
<pre>
[css-scroll-snap] focus should not take into account scroll margin for visibility check
<a href="https://bugs.webkit.org/show_bug.cgi?id=245100">https://bugs.webkit.org/show_bug.cgi?id=245100</a>
&lt;rdar://99840839&gt;

Reviewed by Simon Fraser.

When an element is focused, we perform a visibility calculation to determine if we need to scroll
the element into view. This visibility calculation should not take into account the scroll margin of
the element. This prevents a bug where we wont scroll an element into view that is offscreen because
the rect including its scroll margnin is within the viewport.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/scroll-margin-visibility-check.html:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::scrollToFocusedElementInternal):

Canonical link: <a href="https://commits.webkit.org/254732@main">https://commits.webkit.org/254732@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6e4192a39dd34f436321160013eae44a3549d8b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89918 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34463 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20573 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99247 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/156432 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93924 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32954 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28367 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82252 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93568 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26176 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76714 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26092 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81049 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80870 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69097 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30700 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14953 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30437 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15894 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3324 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33899 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38854 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32616 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34978 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2446,7 +2446,7 @@ void FrameView::scrollToFocusedElementInternal()
     bool insideFixed;
     LayoutRect absoluteBounds = renderer->absoluteAnchorRectWithScrollMargin(&insideFixed);
-    FrameView::scrollRectToVisible(absoluteBounds, *renderer, insideFixed, { m_selectionRevealModeForFocusedElement, ScrollAlignment::alignCenterIfNeeded, ScrollAlignment::alignCenterIfNeeded, ShouldAllowCrossOriginScrolling::No });
+    FrameView::scrollRectToVisible(absoluteBounds, *renderer, insideFixed, { m_selectionRevealModeForFocusedElement, ScrollAlignment::alignCenterIfNeeded, ScrollAlignment::alignCenterIfNeeded, ShouldAllowCrossOriginScrolling::No, ScrollBehavior::Auto, renderer->absoluteAnchorRect(&insideFixed) });
```
I prefer using a local variable to hold the result of `renderer->absoluteAnchorRect(&insideFixed)`; it makes it easier to debug, and you can name it.

We're also now calling `absoluteAnchorRectWithScrollMargin` and `absoluteAnchorRect` both of which are going to do ancestor traversals. Maybe we should modify `absoluteAnchorRectWithScrollMargin` to return both rects.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -406,7 +406,7 @@ class ScrollableArea : public CanMakeWeakPtr<ScrollableArea> {
     bool hasLayerForScrollCorner() const;
 
-    LayoutRect getRectToExposeForScrollIntoView(const LayoutRect& visibleRect, const LayoutRect& exposeRect, const ScrollAlignment& alignX, const ScrollAlignment& alignY) const;
+    LayoutRect getRectToExposeForScrollIntoView(const LayoutRect& visibleRect, const LayoutRect& exposeRect, const ScrollAlignment& alignX, const ScrollAlignment& alignY, const LayoutRect& visibilityRect = LayoutRect()) const;
```
It's hard to know what the difference is between `visibleRect` and `visibilityRect` here.

Also to provided a default value for a type with a default constructor you can just say `= { }`

nmoucht replied:
Switched the first to visibleBounds and the other to visibleCheckRect.

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -142,6 +143,7 @@ struct ScrollRectToVisibleOptions {
     const ScrollAlignment& alignY { ScrollAlignment::alignCenterIfNeeded };
     ShouldAllowCrossOriginScrolling shouldAllowCrossOriginScrolling { ShouldAllowCrossOriginScrolling::No };
     ScrollBehavior behavior { ScrollBehavior::Auto };
+    LayoutRect visibilityRect { LayoutRect() };
```
LayoutRect has a default constructor. No need to provide an initializer.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -1814,17 +1811,27 @@ LayoutRect RenderLayerScrollableArea::scrollRectToVisible(const LayoutRect& abso
 
-    auto revealRect = getRectToExposeForScrollIntoView(layerBounds, localExposeRect, options.alignX, options.alignY);
+    LayoutRect localExposeRect = getLocalExposeRect(absoluteRect, box, verticalScrollbarWidth(), layerBounds);
+    LayoutRect localVisiblityRect = getLocalExposeRect(options.visibilityRect, box, verticalScrollbarWidth(), layerBounds);
```
Why do we compute `absoluteRect` and `options.visibilityRect` in absolute coordinates, only to convert them back to local coordinates here? That's error prone and in some cases involving 3d transforms might even not produce the same results.

nmoucht replied:
This is how the code was operating previously. I'm not sure if there's a good way to get the local rect directly without converting the absolute rect.

smfr commented on Source/WebCore/rendering/RenderElement.h:
```diff
@@ -195,7 +195,7 @@ class RenderElement : public RenderObject {
     // absoluteAnchorRectWithScrollMargin() is similar to absoluteAnchorRect, but it also takes into account any
     // CSS scroll-margin that is set in the style of this RenderElement.
-    LayoutRect absoluteAnchorRectWithScrollMargin(bool* insideFixed = nullptr) const;
+    std::pair<LayoutRect, LayoutRect> absoluteAnchorRectWithScrollMargin(bool* insideFixed = nullptr) const;
```
Returning a std::pair here makes the code hard to read; you have to dig around to find out what `.first` and `.second` represent. It would be better to return a struct, either with two rects, or a rect and a margin.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollableArea.cpp:
```diff
@@ -808,23 +808,28 @@ std::optional<BoxSide> ScrollableArea::targetSideForScrollDelta(FloatSize delta,
-
+    
+    bool intersectsInX = false;
+    if (!visibilityCheckRect.isEmpty())
```
We should represent an unset visibilityCheckRect as a `std::optional<LayoutRect>` rather than checking if it's empty.

nmoucht replied:
Fixed.

nmoucht committed:
[css-scroll-snap] focus should not take into account scroll margin for visibility check
https://bugs.webkit.org/show_bug.cgi?id=245100
<rdar://99840839>

Reviewed by Simon Fraser.

When an element is focused, we perform a visibility calculation to determine if we need to scroll
the element into view. This visibility calculation should not take into account the scroll margin of
the element. This prevents a bug where we wont scroll an element into view that is offscreen because
the rect including its scroll margnin is within the viewport.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/scroll-margin-visibility-check.html:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::scrollToFocusedElementInternal):

Canonical link: https://commits.webkit.org/254732@main
